Sources/XCTest/Public/Asynchronous/XCTWaiter.swift

```diff
@@ -187,9 +189,16 @@ open class XCTWaiter {
     ///   these environments. To ensure compatibility of tests between
     ///   swift-corelibs-xctest and Apple XCTest, it is not recommended to pass
     ///   explicit values for `file` and `line`.
+    #if USE_SWIFT_CONCURRENCY_WAITER
+    @available(*, unavailable, message: "Expectation-based waiting is not available when using the Swift concurrency waiter.")
```

Comment:
I'm a little bit confused about the messaging of the `USE_SWIFT_CONCURRENCY_WAITER` option, the continued availability of the `XCTWaiter` class, and these unavailability messages about "expectation-based waiting".

Is there anything left that `XCTWaiter` *can* do in this mode? Or should the entire class be removed / made unavailable?

Reply:
Do you think it would be workable to adjust the scope and semantics of this conditional a little bit, perhaps calling it `DISABLE_XCTWAITER` or something along those lines, and have it guard the entirety of the `XCTWaiter` and `XCTestExpectation` classes, along with the everything in `XCTestCase+Asynchronous.swift`, and the expectation-related state tracking in `XCTestCase.swift`.

I do see that usage of the `XCTWaiter.subsystemQueue` has leaked out to non-waiter code in a few places (`ThrownErrorWrapper` and `TeardownBlocksState`). Those are using it for just basic locking purposes, and quite independently from the actual waiter subsystem. Perhaps we introduce a ` XCTestCase.subsystemQueue` to use for those instead? Or even a different locking solution altogether? This would be a general-purpose change, not conditionalized by the new `USE_SWIFT_CONCURRENCY_WAITER`/`DISABLE_XCTWAITER` option.

Reply:
If we introduce a second queue, we should be very wary of deadlocks, data races, etc. where the current code assumes a single global lock.

Reply:
Right, I used the new queue in `TeardownBlocksState` and `ThrownErrorWrapper`. As far as I checked, they are well independent from the other, and they won't acquire the other lock while acquiring itself lock. But I need more eyes here :pray:

Reply:
Agreed, considering the fine-grained scoping of the locking provided by the queue for these, I think we should be ok in that regard.